libass/ass_render.c

```diff
@@ -2079,6 +2080,24 @@ static void calculate_rotation_params(ASS_Renderer *render_priv, ASS_DRect *bbox
 }
 
 
+static int quantize_blur(double radius, int32_t *shadow_mask)
+{
+    const double scale = TRANFORM_ACCURACY / (64 * BLUR_ACCURACY);
+    radius *= 2 / sqrt(log(256)) / scale;
+
+    int ord;
+    frexp((1 + radius) * (TRANFORM_ACCURACY / 2), &ord);
+    *shadow_mask = ((uint32_t) 1 << ord) - 1;
+    return lrint(log1p(radius) / BLUR_ACCURACY);
+}
+
+static double restore_blur(int qblur)
+{
+    const double scale = TRANFORM_ACCURACY / (64 * BLUR_ACCURACY);
+    double sigma = scale * (exp(BLUR_ACCURACY * qblur) - 1);
```

Comment:
Oh hey. You can use `expm1` here. And it’s guaranteed to return exactly zero for zero argument, too.

Reply:
Er… guaranteed if the platform supports IEEE arithmetic (Annex F of the C standard). Nothing is guaranteed if it doesn’t. So I guess we do need to keep that epsilon over in `ass_synth_blur`.

Reply:
I could easily imagine people are trying to build libass with stuff like -ffast-math. Also clang apparently does not formally support IEEE semantics. On the other hand, "random" float semantics are just crazy and you can't write working code for that.

Reply:
I'm building libass with -ffast-math. On the other hand I consider writing under "random" float semantics (result of every FP operation have random error in the range of couple ULPs) is the only right way to write math. Relying on strict semantics make program fragile and can lead to numerical instabilities.